File: cogutils_reservoirpy/reservoir_node.py
"""ReservoirPy Node implementation for OpenCog."""

from typing import Optional, Union, Dict, Any
import numpy as np
from reservoirpy.nodes import Node
from atomspace import AtomSpace, Atom, types

class ReservoirNode(Node):
    """ReservoirPy Node that can interface with OpenCog AtomSpace.
    
    Parameters
    ----------
    units : int
        Number of reservoir units
    input_scaling : float
        Input scaling factor
    spectral_radius : float
        Spectral radius of reservoir
    atomspace : AtomSpace
        OpenCog AtomSpace instance
    """
    
    def __init__(
        self,
        units: int = 100,
        input_scaling: float = 1.0,
        spectral_radius: float = 0.99,
        atomspace: Optional[AtomSpace] = None,
        **kwargs
    ):
        super().__init__(
            params={
                "W": None,  # Reservoir weights
                "Win": None,  # Input weights
                "state": None,  # Current state
            },
            hypers={
                "units": units,
                "input_scaling": input_scaling,
                "spectral_radius": spectral_radius,
            },
            **kwargs
        )
        self.atomspace = atomspace or AtomSpace()
        
    def _initialize(self, x: np.ndarray = None) -> None:
        """Initialize reservoir weights and state."""
        n_inputs = x.shape[1] if x is not None else 1
        n_units = self.hypers["units"]
        
        # Initialize input weights
        self.params["Win"] = (
            np.random.uniform(-1, 1, (n_units, n_inputs)) * 
            self.hypers["input_scaling"]
        )
        
        # Initialize reservoir weights
        W = np.random.uniform(-1, 1, (n_units, n_units))
        radius = np.max(np.abs(np.linalg.eigvals(W)))
        self.params["W"] = W * (self.hypers["spectral_radius"] / radius)
        
        # Initialize state
        self.params["state"] = np.zeros((1, n_units))
        
    def forward(self, x: np.ndarray) -> np.ndarray:
        """Compute reservoir state update.
        
        Parameters
        ----------
        x : np.ndarray
            Input data
            
        Returns
        -------
        np.ndarray
            Updated reservoir state
        """
        if self.params["W"] is None:
            self._initialize(x)
            
        prev_state = self.params["state"]
        
        # Update reservoir state
        new_state = np.tanh(
            np.dot(x, self.params["Win"].T) +
            np.dot(prev_state, self.params["W"].T)
        )
        
        self.params["state"] = new_state
        return new_state
        
    def to_atoms(self) -> Dict[str, Atom]:
        """Convert reservoir state to Atoms in AtomSpace.
        
        Returns
        -------
        Dict[str, Atom]
            Dictionary mapping parameter names to Atoms
        """
        atoms = {}
        for name, param in self.params.items():
            if param is not None:
                # Create NumberNode for each parameter value
                atom = self.atomspace.add_node(
                    types.NumberNode,
                    str(param.flatten()[0])
                )
                atoms[name] = atom
        return atoms

File: src/ReservoirNode.ts
import { Matrix, multiply, random, tanh } from 'mathjs';
import { AtomSpace, Node as AtomNode } from '@opencog/atomspace';

export class ReservoirNode {
  private units: number;
  private inputScaling: number;
  private spectralRadius: number;
  private atomspace: AtomSpace;
  
  private Win?: Matrix;  // Input weights
  private W?: Matrix;    // Reservoir weights
  private state?: Matrix; // Current state

  constructor(
    units: number = 100,
    inputScaling: number = 1.0,
    spectralRadius: number = 0.99,
    atomspace?: AtomSpace
  ) {
    this.units = units;
    this.inputScaling = inputScaling;
    this.spectralRadius = spectralRadius;
    this.atomspace = atomspace || new AtomSpace();
  }

  private initialize(inputDim: number): void {
    // Initialize input weights
    this.Win = multiply(
      random([this.units, inputDim], -1, 1),
      this.inputScaling
    ) as Matrix;

    // Initialize reservoir weights with spectral radius scaling
    const W = random([this.units, this.units], -1, 1) as Matrix;
    const eigenvalues = W.eigs().values;
    const maxEigenvalue = Math.max(...eigenvalues.map(Math.abs));
    this.W = multiply(W, this.spectralRadius / maxEigenvalue) as Matrix;

    // Initialize state
    this.state = random([1, this.units], 0, 0) as Matrix;
  }

  forward(input: Matrix): Matrix {
    if (!this.W || !this.Win || !this.state) {
      this.initialize(input.size()[1]);
    }

    // Update reservoir state
    const inputProjection = multiply(input, this.Win!.transpose());
    const recurrentProjection = multiply(this.state!, this.W!);
    this.state = tanh(add(inputProjection, recurrentProjection)) as Matrix;

    return this.state;
  }

  toAtoms(): Map<string, AtomNode> {
    const atoms = new Map<string, AtomNode>();

    if (this.state) {
      // Convert state to NumberNode atoms
      const stateValues = this.state.valueOf() as number[][];
      stateValues.forEach((row, i) => {
        row.forEach((value, j) => {
          const atom = this.atomspace.addNode(
            'NumberNode',
            `state_${i}_${j}`,
            value
          );
          atoms.set(`state_${i}_${j}`, atom);
        });
      });
    }

    return atoms;
  }

  getState(): Matrix | undefined {
    return this.state;
  }

  reset(): void {
    if (this.state) {
      this.state = random([1, this.units], 0, 0) as Matrix;
    }
  }
}